src/main/java/com/epam/aidial/core/storage/BlobStorage.java

```diff
@@ -0,0 +1,199 @@
+        List<FileMetadataBase> metadata = new ArrayList<>();
+        ListContainerOptions options = buildListContainerOptions(BlobStorageUtil.normalizePathForQuery(path));
+        PageSet<? extends StorageMetadata> list = blobStore.list(bucketName, options);
+        list.forEach(meta -> {
+            StorageType objectType = meta.getType();
+            switch (objectType) {
+                case BLOB -> metadata.add(buildFileMetadata(meta));
+                case FOLDER, RELATIVE_PATH -> metadata.add(buildFolderMetadata(meta));
+                default -> throw new IllegalArgumentException("Can't list container");
+            }
+        });
+
+        return metadata;
+    }
+
+    public void close() {
+        storeContext.close();
+    }
+
+    private static ListContainerOptions buildListContainerOptions(String prefix) {
+        return new ListContainerOptions()
+                .prefix(prefix)
+                .delimiter(BlobStorageUtil.PATH_SEPARATOR);
+    }
+
+    private static FileMetadata buildFileMetadata(StorageMetadata metadata) {
+        String absoluteFilePath = metadata.getName();
+        String[] elements = absoluteFilePath.split(BlobStorageUtil.PATH_SEPARATOR);
+        String fileName = elements[elements.length - 1];
+        String path = absoluteFilePath.substring(0, absoluteFilePath.length() - fileName.length() - 1);
+        return new FileMetadata(fileName, path, metadata.getSize(), BlobStorageUtil.getContentType(fileName));
+    }
+
+    private static FolderMetadata buildFolderMetadata(StorageMetadata metadata) {
+        String absoluteFolderPath = metadata.getName();
+        String[] elements = absoluteFolderPath.split(BlobStorageUtil.PATH_SEPARATOR);
+        String lastFolderName = elements[elements.length - 1];
+        String path = absoluteFolderPath.substring(0, absoluteFolderPath.length() - lastFolderName.length() - 1);
+        return new FolderMetadata(BlobStorageUtil.removeLeadingAndTrailingPathSeparators(lastFolderName),
+                BlobStorageUtil.removeTrailingPathSeparator(path));
+    }
+
+    private static BlobMetadata buildBlobMetadata(String fileName, String path, String contentType, String bucketName) {
+        String filePath = BlobStorageUtil.buildFilePath(fileName, path);
+        ContentMetadata contentMetadata = buildContentMetadata(contentType);
+        return new BlobMetadataImpl(null, filePath, null, null, null, null, null, Map.of(), null, bucketName, contentMetadata, null, Tier.STANDARD);
+    }
+
+    private static ContentMetadata buildContentMetadata(String contentType) {
+        ContentMetadata contentMetadata = ContentMetadataBuilder.create()
+                .contentType(contentType)
+                .build();
+        return BaseMutableContentMetadata.fromContentMetadata(contentMetadata);
+    }
+
+    private void createBucketIfNeeded(Storage config) {
+        if (config.getProvider().equals("google-cloud-storage")) {
```

Comment:
It's ok to fail the operation on createBucket if user explicitly set the flag `needToCreateBucket` to `true` and the account doesn't have permissions. That's an issue with permissions on provider account level.
I think we should add a config property `needToCreateBucket` as a boolean value.

Reply:
fixed